<a href="https://colab.research.google.com/github/VidishaPonnappa/MachineLearning/blob/main/CapstoneProject_7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project 7: Bengaluru House Prices - Data Cleaning


---

### Context

A techie residing in Mumbai who is a data scientist got an excellent job opportunity from one of his dream company in Bengaluru. Now, he wants to shift to Bengaluru soon and decides to buy a home in the city.
The lack of trust in property developers in the city resulted in a drop of 7% in the housing units sales across India in 2017. The property prices in Bengaluru fell by almost 5% in the second half of 2017, said a study published by a property consultancy firm called Knight Frank. Here's the link to the news article:

[Residential real estate prices drop 3% in big cities
](https://www.businesstoday.in/sectors/infra/residential-real-estate-prices-property-drop-3-per-cent-realty/story/267752.html)

Buying a home, especially in a city like Bengaluru, is a tricky choice. While the major factors are usually the same for all metro cities, there are others to be considered for the Silicon Valley of India. Because of the millennial crowd, vibrant culture, great climate and a slew of job opportunities in Bengaluru, it is difficult to ascertain the price of a house in the city.
As he is a data scientist he finds the dataset for Bengaluru house prices on the internet.

The dataset he found was vast with some useless information and some of the empty values, he only wants to extract some of the useful information from the dataset.

---

### Problem Statement

The dataset acquired by the techie is full of irregularities, incorrect values and missing values. As a data scientist (or analyst in this context), your task is to clean the dataset given to you in this problem statement.

This process of preparing data for analysis by removing or modifying data that is incorrect, incomplete, irrelevant, duplicated, or improperly formatted is known as data cleaning.


---

### Dataset Information

This dataset contains several factors that a potential home buyer considers before purchasing a house. They are the following:

1. `area_type`: Whether the house has a built-up area or plot area.

2. `availability`: Whether the house is in the ready-to-move state or the exact date on which the house will be ready for occupancy.

3. `location`: The name of the locality in Bengaluru where the house is situated.

4. `size`: Whether the house is 1 BHK, 2 BHK, 3 BHK etc. The term BHK stands for *Bedroom, Hall & Kitchen*.

5. `society`: Name of the society where the house is located.

6. `total_sqft`: The total area of the house in square feet.

7. `bath`: Number of bathrooms in the house.

8. `balcony`: Number of balconies in the house.

9. `price`: Price of the house in lakhs (Indian rupees).


---

### Things To Do

- Create a Pandas DataFrame for the Bengaluru house prices dataset.

- Find the missing values in the DataFrame and replace them with the right missing values. However, if a column has more than 15% missing values then drop (or remove) the column from the DataFrame except for the `location, size` and `total_sqft` columns because the house prices are directly dependent on these three crucial parameters. *In fact, the rate of a house is reported as a cost per unit area of the house*.

- Drop (or remove) the rows containing the missing values in the `location,	size` and `total_sqft` columns. Additionally, drop any row which contains nonsensical values in the context of houses.

---

#### 1. Import Modules

First import all the necessary modules to create a DataFrame and box plots.

In [ ]:
# Import the necessary modules.
import pandas as pd

---

#### 2. Load The Dataset

Create a Pandas DataFrame by loading the dataset which is provided in the comma-separated values (CSV) file. Also, display the number of rows and columns in the DataFrame. Here's the link to the dataset:

https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/bengaluru-house-prices/Bengaluru_House_Prices.csv

In [ ]:
# Load the dataset.
df=pd.read_csv('https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/bengaluru-house-prices/Bengaluru_House_Prices.csv')
df

area_type   availability  ... balcony   price
0      Super built-up  Area         19-Dec  ...     1.0   39.07
1                Plot  Area  Ready To Move  ...     3.0  120.00
2            Built-up  Area  Ready To Move  ...     3.0   62.00
3      Super built-up  Area  Ready To Move  ...     1.0   95.00
4      Super built-up  Area  Ready To Move  ...     1.0   51.00
...                     ...            ...  ...     ...     ...
13315        Built-up  Area  Ready To Move  ...     0.0  231.00
13316  Super built-up  Area  Ready To Move  ...     NaN  400.00
13317        Built-up  Area  Ready To Move  ...     1.0   60.00
13318  Super built-up  Area         18-Jun  ...     1.0  488.00
13319  Super built-up  Area  Ready To Move  ...     1.0   17.00

[13320 rows x 9 columns]

In [ ]:
# Display the number of rows and columns.
df.shape

(13320, 9)

---

#### 3. Check For The Missing Values

- Compute the total number of missing values in the DataFrame.

- Compute the percentage of missing values in the DataFrame. If a column has more than 15% missing values then drop (or remove) the column from the DataFrame.

In [ ]:
# Find the total number of missing values in each column.
df.isnull().sum()

area_type          0
availability       0
location           1
size              16
society         5502
total_sqft         0
bath              73
balcony          609
price              0
dtype: int64

**Hint**: You can find the total number of missing values in each column by using the `sum()` function on top of the `isnull()` function.

**Q:** Which column has the most number of missing values?

**A:** society

In [ ]:
# Find the percentage of missing values
df.isnull().sum()*100/df.shape[0]

area_type        0.000000
availability     0.000000
location         0.007508
size             0.120120
society         41.306306
total_sqft       0.000000
bath             0.548048
balcony          4.572072
price            0.000000
dtype: float64

**Hint** : You can calculate the percentage of missing values in each column by multiplying the total number of missing values in each column with 100 and dividing the value with the total number of rows in the DataFrame.

**Q:** Which is/are column(s) that need to be removed? Write "Not Applicable" if there are none.

**A:** society

---

### 4. Data Cleaning - Rows

First things first, drop (or remove) the rows containing the missing values in the `location,	size` and `total_sqft` columns because they are the most important parameters in governing the price of a house.

---

#### 4.1 Removing Rows From The `location` Column

You need to remove all the rows containing the missing values in the `location` column.

In [ ]:
# Get all the rows having the missing values in the 'location'.
#ml=df[df['location'].isnull()=='True']
ml=df.loc[df['location'].isnull()==True,:]
ml

area_type   availability location  ... bath balcony price
568  Super built-up  Area  Ready To Move      NaN  ...  3.0     2.0  86.0

[1 rows x 9 columns]

**Q:** How many rows have the missing values in the `location` column?

**A:** 1

In [ ]:
# Total number of rows having the missing values in the 'location' column.
ml['location'].shape

(13319,)

In [ ]:
# Discard the rows containing the missing values in the 'location' column.
df=df.loc[df['location'].isnull()==False,:]
df

area_type   availability  ... balcony   price
0      Super built-up  Area         19-Dec  ...     1.0   39.07
1                Plot  Area  Ready To Move  ...     3.0  120.00
2            Built-up  Area  Ready To Move  ...     3.0   62.00
3      Super built-up  Area  Ready To Move  ...     1.0   95.00
4      Super built-up  Area  Ready To Move  ...     1.0   51.00
...                     ...            ...  ...     ...     ...
13315        Built-up  Area  Ready To Move  ...     0.0  231.00
13316  Super built-up  Area  Ready To Move  ...     NaN  400.00
13317        Built-up  Area  Ready To Move  ...     1.0   60.00
13318  Super built-up  Area         18-Jun  ...     1.0  488.00
13319  Super built-up  Area  Ready To Move  ...     1.0   17.00

[13303 rows x 9 columns]

**Hint** : Use `isnull()` function inside the `loc[]` function to locate and remove the rows with missing values from the `location` column.

---

#### 4.2 Removing Rows From The `size` Column

Now, you need to remove all the rows containing the missing values in the `size` column.

In [ ]:
# Get the rows having the missing values in the 'size' column.
sizem=df.loc[df['size'].isnull()==True,:]
sizem

area_type          availability        location  ... bath balcony    price
579    Plot  Area  Immediate Possession  Sarjapur  Road  ...  NaN     NaN   34.185
1775   Plot  Area  Immediate Possession        IVC Road  ...  NaN     NaN  124.000
2264   Plot  Area  Immediate Possession    Banashankari  ...  NaN     NaN  460.000
2809   Plot  Area  Immediate Possession  Sarjapur  Road  ...  NaN     NaN   28.785
2862   Plot  Area  Immediate Possession     Devanahalli  ...  NaN     NaN   46.800
5333   Plot  Area  Immediate Possession     Devanahalli  ...  NaN     NaN  177.115
6423   Plot  Area  Immediate Possession      Whitefield  ...  NaN     NaN   26.730
6636   Plot  Area  Immediate Possession          Jigani  ...  NaN     NaN   25.490
6719   Plot  Area  Immediate Possession         Hoskote  ...  NaN     NaN   28.545
7680   Plot  Area  Immediate Possession    Kasavanhalli  ...  NaN     NaN  400.000
8306   Plot  Area  Immediate Possession          Anekal  ...  NaN     NaN   16.500
8565   Plot  Area  Immediate Possession      Carmelaram  ...  NaN     NaN  120.000
8703   Plot  Area  Immediate Possession     Mysore Road  ...  NaN     NaN   42.300
10634  Plot  Area  Immediate Possession      Rajankunte  ...  NaN     NaN   31.110
11019  Plot  Area  Immediate Possession       Yelahanka  ...  NaN     NaN   12.750
11569  Plot  Area  Immediate Possession      Hosur Road  ...  NaN     NaN    8.440

[16 rows x 9 columns]

**Q:** How many rows have the missing values in the `size` column?

**A:** 16

In [ ]:
# Total number of rows having the missing values in the 'size' column.
sizem.shape

(16, 9)

In [ ]:
# Discard the rows containing the missing values in the 'size' column.
df=df.loc[df['size'].isnull()==False,:]
df

area_type   availability  ... balcony   price
0      Super built-up  Area         19-Dec  ...     1.0   39.07
1                Plot  Area  Ready To Move  ...     3.0  120.00
2            Built-up  Area  Ready To Move  ...     3.0   62.00
3      Super built-up  Area  Ready To Move  ...     1.0   95.00
4      Super built-up  Area  Ready To Move  ...     1.0   51.00
...                     ...            ...  ...     ...     ...
13315        Built-up  Area  Ready To Move  ...     0.0  231.00
13316  Super built-up  Area  Ready To Move  ...     NaN  400.00
13317        Built-up  Area  Ready To Move  ...     1.0   60.00
13318  Super built-up  Area         18-Jun  ...     1.0  488.00
13319  Super built-up  Area  Ready To Move  ...     1.0   17.00

[13303 rows x 9 columns]

**Q:** How many rows are there in the DataFrame after removing the rows that need to be removed?

**A:**

---

#### 4.3 Removing Rows From The `total_sqft` Column

Now, you need to remove all the rows containing the missing values in the `total_sqft` column.

In [ ]:
# Get the rows having the missing values in the 'total_sqft' column.
tsqm=df.loc[df['total_sqft'].isnull()==True,:]
tsqm

Empty DataFrame
Columns: [area_type, availability, location, size, society, total_sqft, bath, balcony, price]
Index: []

**Q:** How many rows have the missing values in the `size` column?

**A:** 0

---

#### 4.3 Removing Rows From The `bath` Column

Assume that in the practical situation, no house, even in a city like Bengaluru, will have more than 5 bathrooms. Hence, drop all the rows in the `bath` column containing more than 5 bathrooms.

In [ ]:
# Get the rows having more than 5 bathrooms in the 'bath' column.
bathexcess=df.loc[df['bath']>5,:]
bathexcess

area_type   availability  ... balcony  price
9          Plot  Area  Ready To Move  ...     NaN  370.0
45         Plot  Area  Ready To Move  ...     NaN  200.0
64     Built-up  Area  Ready To Move  ...     0.0  140.0
68         Plot  Area  Ready To Move  ...     0.0   85.0
79     Built-up  Area  Ready To Move  ...     3.0  250.0
...               ...            ...  ...     ...    ...
13221      Plot  Area  Ready To Move  ...     1.0   75.0
13226  Built-up  Area  Ready To Move  ...     3.0  215.0
13277      Plot  Area  Ready To Move  ...     NaN  218.0
13300      Plot  Area  Ready To Move  ...     2.0  145.0
13311      Plot  Area  Ready To Move  ...     2.0  250.0

[515 rows x 9 columns]

**Q:** How many rows have more than 5 bathrooms in the `bath` column?

**A:**

In [ ]:
# Discard the rows having more than 5 bathrooms in the 'bath' column.
df=df.loc[df['bath']<=5]
df

area_type   availability  ... balcony   price
0      Super built-up  Area         19-Dec  ...     1.0   39.07
1                Plot  Area  Ready To Move  ...     3.0  120.00
2            Built-up  Area  Ready To Move  ...     3.0   62.00
3      Super built-up  Area  Ready To Move  ...     1.0   95.00
4      Super built-up  Area  Ready To Move  ...     1.0   51.00
...                     ...            ...  ...     ...     ...
13315        Built-up  Area  Ready To Move  ...     0.0  231.00
13316  Super built-up  Area  Ready To Move  ...     NaN  400.00
13317        Built-up  Area  Ready To Move  ...     1.0   60.00
13318  Super built-up  Area         18-Jun  ...     1.0  488.00
13319  Super built-up  Area  Ready To Move  ...     1.0   17.00

[12731 rows x 9 columns]

**Q:** How many rows have more at most 5 bathrooms in the `bath` column?

**A:** 12731

---

### 5. Data Cleaning - Columns

Compute the percentage of missing values in the DataFrame again. If a column has more than 15% missing values then drop (or remove) the column from the DataFrame.


In [ ]:
# Percentage of missing values.
df.isnull().sum()*100/df.shape[0]

area_type        0.000000
availability     0.000000
location         0.000000
size             0.000000
society         39.910455
total_sqft       0.000000
bath             0.000000
balcony          3.361873
price            0.000000
dtype: float64

**Hint**: You can calculate the percentage of missing values in each column by multiplying the number of missing values with 100 and dividing the resulting value with total number of rows in the DataFrame.

**Q:** Which is/are column(s) that need to be removed? Write "Not Applicable" if there are none.

**A:** society

**Q:** Did the overall percentage of the missing values dropped compared to the initial percentage of the missing values? Write yes or no.

**A:** Yes

---

#### 5.1 Columns To Be Retained

Create a list of columns that need to be retained.

In [ ]:
# List of the columns to be retained.
retcol=[]
for i in df.columns:
  if i!='society':
    retcol.append(i)
retcol

['area_type',
 'availability',
 'location',
 'size',
 'total_sqft',
 'bath',
 'balcony',
 'price']

In [ ]:
df.columns

Index(['area_type', 'availability', 'location', 'size', 'society',
       'total_sqft', 'bath', 'balcony', 'price'],
      dtype='object')

**Hint**: You need to create a list of columns that have missing values not more than 15%. This can be done by appending such column names in a list using a `for` loop.

---

#### 5.2 Retaining Columns

Retain the appropriate columns in the DataFrame.

In [ ]:
# Retain the appropriate columns in the DataFrame.
df=df.loc[:,retcol]
df

area_type   availability  ... balcony   price
0      Super built-up  Area         19-Dec  ...     1.0   39.07
1                Plot  Area  Ready To Move  ...     3.0  120.00
2            Built-up  Area  Ready To Move  ...     3.0   62.00
3      Super built-up  Area  Ready To Move  ...     1.0   95.00
4      Super built-up  Area  Ready To Move  ...     1.0   51.00
...                     ...            ...  ...     ...     ...
13315        Built-up  Area  Ready To Move  ...     0.0  231.00
13316  Super built-up  Area  Ready To Move  ...     NaN  400.00
13317        Built-up  Area  Ready To Move  ...     1.0   60.00
13318  Super built-up  Area         18-Jun  ...     1.0  488.00
13319  Super built-up  Area  Ready To Move  ...     1.0   17.00

[12731 rows x 8 columns]

**Hint**: You need to create a DataFrame with columns having missing values not more than 15%. This can be done by passing the variable holding the list of columns inside a `loc[]` function.

Again calculate the percetange of missing values in the DataFrame.

In [ ]:
# Percentage of the missing values in the DataFrame.
df.isnull().sum()*100/df.shape[0]

area_type       0.000000
availability    0.000000
location        0.000000
size            0.000000
total_sqft      0.000000
bath            0.000000
balcony         3.361873
price           0.000000
dtype: float64

**Q:** Write down (or identify) the column(s) in which the values need to be replaced with an appropriate value.

**A:** balcony

---